File: models/ApiShow.js
import ApiEntity from "./ApiEntity.js";

class ApiShowPreview extends ApiEntity {

    constructor(show) {
        super(show.id, show.title);
        this.poster = show.poster;
        this.creation = show.release_date;
        this.platforms = getPlatforms(show.svods);
    }
}

class ApiShow extends ApiEntity {

    /**
     * @param {Object} show 
     */
    constructor(show) {
        super(show.id, show.title);
        this.poster = getImageUrl(show.images);
        this.duration = show.length ?? 0;
        this.country = show.country;
        this.description = show.description;
        this.seasons = show.seasons;
        this.episodes = show.episodes;
        this.network = show.network;
        this.note = getNote(show.notes);
        this.status = show.status === "Continuing" ? "En cours" : "Terminée";
        this.creation = show.creation;
        this.kinds = Object.values(show.genres);
        this.platforms = getPlatforms(show.platforms?.svods);
    }
}

/**
 * @param {object} platforms 
 * @return object[]
 */
const getPlatforms = (platforms) => platforms ? platforms.map((p) => (
    {
        "name": p.name,
        "logo": p.logo
    }
)) : [];

/**
 * @param {Object?} note 
 * @returns number
 */
const getNote = (note) => {
    if (Object.keys(note ?? {}).length === 0) return null;
    return note.mean;

}

/**
 * @param {Object?} image 
 * @returns string
 */
const getImageUrl = (image) => {
    if (Object.keys(image ?? {}).length === 0) return null;
    if (image.poster) return image.poster;
    if (image.show) return image.show;
    if (image.banner) return image.banner;
    if (image.box) return image.box;
    return null;
}

export { ApiShow, ApiShowPreview };

File: models/api_show.go
package models

type DiscoverApiShows struct {
	Shows []ApiShow `json:"shows"`
}

type DisplayApiShow struct {
	Show ApiShow `json:"show"`
}

type ApiShow struct {
	Id     uint   `json:"id"`
	Title  string `json:"title"`
	Images struct {
		Banner string `json:"banner,omitempty"`
		Box    string `json:"box,omitempty"`
		Poster string `json:"poster,omitempty"`
		Show   string `json:"show,omitempty"`
	} `json:"images"`
	Description string `json:"description"`
	Episodes    string `json:"episodes"`
	Seasons     []struct {
		Number   uint `json:"number"`
		Episodes uint `json:"episodes"`
	} `json:"seasons_details"`
	Creation  string      `json:"creation"`
	Country   string      `json:"country"`
	Genres    interface{} `json:"genres"`
	Duration  uint        `json:"length"`
	Status    string      `json:"status"`
	Notes     interface{} `json:"notes"`
	Platforms struct {
		Streaming []struct {
			Name string `json:"name"`
			Logo string `json:"logo"`
		} `json:"svods"`
	} `json:"platforms"`
}

func (show ApiShow) GetImageUrl() string {
	if show.Images.Poster != "" {
		return show.Images.Poster
	}
	if show.Images.Show != "" {
		return show.Images.Show
	}
	if show.Images.Banner != "" {
		return show.Images.Banner
	}
	if show.Images.Box != "" {
		return show.Images.Box
	}
	return ""
}
